lib/src/front_end/piece_writer.dart

```diff
@@ -76,19 +133,31 @@ class PieceWriter {
   PieceWriter(this._formatter, this._source);
 
   /// Gives the builder a newly completed [piece], to be taken by a later call
-  /// to [pop] from some surrounding piece.
+  /// to [pop()] or [split()] from some surrounding piece.
   void push(Piece piece) {
```

Comment:
I wonder if there are any other names that would convey that this is a single slot. Maybe something like `offer` and `take`?

The `push` and `pop` are too engrained with collections, and I'm finding it hard to reason about this CL with those wrong interpretations so accessible. Some of the code changes look like bugs if it behaved like a collection.

Reply:
Ugh, yeah. I've agonized over what to call them. I did have `give()` and `take()` for a while. I tried using `end()` for `pop()`, but that felt weird because there wasn't really a corresponding `begin()`. I considered calling `split()` `begin()`, but then it's weird that it returns the *previous* piece and not the piece it just begun.

Maybe `send()` and `receive()`?

Naming is hard. @kallentu, what do you think of `give()` for `push()` and `take()` for `pop()`?

Reply:
I thought Nate's `shelve` and `unshelve` kinda made sense. It does almost feel like staging a piece for later.

I don't know how I feel about `give` and `take`. I feel like `push` and `pop` almost reads better than those.

Reply:
`shelve()` and `unshelve()` feel weird for me because it sounds like you're putting the piece aside for some arbitrary amount of time when the shelved piece really will be the next one yielded up by the PieceWriter (which is why I initially went with `push()` and `pop()`).

Maybe `add()` for `push()` and `complete()` for `pop()`?

Reply:
An additional thought: I was thinking of `split()` and I was wondering if we could name `split` in a way that references `pop`. It's not clear immediately that `split` also `pop`s. 

`give()` `take()` `takeAndBegin()` => or something? I couldn't make a better example with `add()` and `complete()`. Take with a grain of salt I suppose haha. I just thought it'd help me read the code better.

Naming is hard :/

Reply:
OK, I spent some more time mulling over this and trying out a few options. Looking at the callsites and what reads most naturally in the docs, I think `give()`/`take()` are my favorites. I went ahead and changed it to that.

We can always iterate on it later if it doesn't grow on us. But I think this PR overall improves things by having `split()` be a combined operation for splitting and returning the previous piece.